libs/utils/trace_analysis.py

```diff
@@ -785,3 +787,117 @@ def plotSchedTuneConf(self):
         # Save generated plots into datadir
         figname = '{}/{}schedtune_conf.png'.format(self.plotsdir, self.prefix)
         pl.savefig(figname, bbox_inches='tight')
+
+    def getClusterResidency(self, cluster, cluster_name):
+        """
+        Get a DataFrame with per cluster frequency residency, i.e. amount of
+        time spent at a given frequency in each cluster.
+
+        :param cluster: list of CPU IDs belonging to the cluster
+        :type cluster: list(int)
+
+        :param cluster_name: name of the cluster
+        :type cluster_name: str
+        """
+
+        if not self.trace.hasEvents('cpu_frequency'):
+            logging.warn('Events [cpu_frequency] not found, '\
+                         'plot DISABLED!')
+            return None
+        if not self.trace.hasEvents('cpu_idle'):
+            logging.warn('Events [cpu_idle] not found, '\
+                         'plot DISABLED!')
+            return None
+
+        freq_dfr = self.trace.df('cpu_frequency')
+        cpu_freqs = freq_dfr[freq_dfr.apply(lambda x: x.cpu in cluster,
+                                            axis=1)]
+        idle_dfr = self.trace.df('cpu_idle')
+        cpu_idle = idle_dfr[idle_dfr.apply(lambda x: x.cpu_id in cluster,
+                                           axis=1)]
+
+        # time instants where CPUs go to an idle state
+        idle_times = cpu_idle[cpu_idle['state'] != IDLE_STATE]
+        # time instants where CPUs leave an idle state
+        non_idle_times = cpu_idle[cpu_idle['state'] == IDLE_STATE]
+
+        if idle_times.index[0] < non_idle_times.index[0]:
+            idle_times = idle_times.iloc[1:]
+
+        time_intervals = [t2 - t1 for t1, t2 in zip(non_idle_times.index,
+                                                    idle_times.index)]
```

Comment:
You can't substract like this because you may substract a time when one cpu goes out of idle from when another cpu went idle.  For example, in the following trace:
```
0.016381 	<idle> 	3 	0 	3 	4294967295
0.016424 	<idle> 	3 	0 	3 	0
0.017242 	<idle> 	2 	0 	2 	1
0.017372 	<idle> 	3 	0 	3 	4294967295
0.017544 	<idle> 	3 	0 	3 	0
0.017618 	<idle> 	3 	0 	3 	4294967295
0.017754 	<idle> 	3 	0 	3 	0
0.017870 	<idle> 	2 	0 	2 	4294967295
0.018141 	<idle> 	3 	0 	3 	4294967295
```
give you the following `time_intervals`:
```
[4.3000000005122274e-05,
 -0.00013000000035390258,
 -7.3999999585794285e-05,
 -0.0001159999992523808]
```

We have to split the `non_indle_times` and `idle_times` into per cpu arrays before doing the substraction. The previous four lines should be something like:
``` python
        idle_non_idle = [(v.index, v2.index) for (_, v), (_, v2) in
                         zip(idle_times.groupby("cpu_id"), non_idle_times.groupby("cpu_id"))]

        time_intervals = []
        for (idle, non_idle) in idle_non_idle:
            if idle[0] < non_idle[0]:
                idle = idle[1:]

            time_intervals.extend([t2 - t1 for t1, t2 in zip(non_idle, idle)])
```